WIP: Create exercises skeleton, take 2: the electric bugaloo

tristan957 opened the pull request:
- [x] Create config tests
- [x] Update documentation
- [x] Create db schema
- [ ] Create initial db models
- [ ] Create `GET /exercises/:id`
- [ ] Create `POST/GET /exercises`
- [ ] Create Neo4j tests with seeded data

tristan957 committed:
Add exercises and exercises-db services

tristan957 committed:
Update README for new service docs

tristan957 committed:
Add requirement.txt

tristan957 commented:
Disregard build failures. It is expecting python files, but I have not committed any due to their unfinished nature at the moment

JosephMart commented on README.md:
```diff
@@ -8,9 +8,16 @@ Contains the source code for the thesis completed under Dr. Philip B. Yasskin
+
+## Documentation
+
+* [Neo4j Python Driver](https://neo4j.com/docs/api/python-driver/current/)
```
what?

tristan957 replied:
Bookmarks to documentation!

JosephMart replied:
I don't think that is readme worthy. Add it to the drive under resources I guess if you really want

tristan957 replied:
SAD! Removed in https://github.com/JosephMart/exercise-dependency-web/pull/2/commits/bc9a007e27fb10056d02bb1ec2c472120b27401a

JosephMart replied:
https://github.com/JosephMart/exercise-dependency-web/pull/2/commits/bc9a007e27fb10056d02bb1ec2c472120b27401a

JosephMart commented on docker-compose-dev.yml:
```diff
@@ -41,3 +41,34 @@ services:
+    container_name: exercises-db
+    image: neo4j:3.5
+    ports:
+      - 7474:7474 # http
```
map these different than source machine i.e `7475:7474`

tristan957 committed:
Make entrypoint scripts executable

tristan957 committed:
Remap Neo4j ports

tristan957 committed:
Remove documentation

JosephMart committed:
Merge branch 'master' into create-exercises-skeleton

JosephMart commented on docker-compose-dev.yml:
```diff
@@ -62,3 +62,34 @@ services:
+      - exercises-db
+
+  exercises-db:
+    container_name: exercises-db
```
no

JosephMart commented on docker-compose-dev.yml:
```diff
@@ -62,3 +62,34 @@ services:
+
+  exercises-db:
+    container_name: exercises-db
+    image: neo4j:3.5
```
no

JosephMart committed:
Finished exercises skeleton

JosephMart committed:
Added API endpoint skeleton

tristan957 committed:
Delete unnecessary config options

tristan957 committed:
Initial creation of the User model

I have defined an abstract base class called Entity. User subclasses
Entity to provide its own serialization and deserialization methods.

tristan957 committed:
Add secrey key back

tristan957 committed:
Add typing to User.serialize and deserialize

tristan957 commented:
User model is done for the most part. Left are all other models.

tristan957 committed:
Include default args

created_at and updated_at should not require someone to pass in the time

tristan957 committed:
Update comments and types

tristan957 committed:
Update types

ID could be optional if we are constructing a new object

tristan957 committed:
Add Chapter model

tristan957 committed:
Add Exercise model

tristan957 committed:
Add Textbook model

tristan957 committed:
Add Topic model

tristan957 committed:
Serialize the ID

tristan957 commented:
What is done is basically model properties. I am sure more will be added as time goes on, and we think about them more. They do not currently have any relationships whatsoever. This is where my knowledge is lacking in Neo4j. Tomorrow I am going to create a test user, and 2 textbooks they have written. Using that data I want to see how the JSON gets returned to me using different Cypher queries. I imagine it will hopefully be something similar to
```json
{
  "id": 1,
  "user_id": 1,
  ...,
  "textbooks": [
    {
      "id": 2,
      "title": "MYMA"
    }
  ]
}

tristan957 committed:
Close DB on app teardown

tristan957 committed:
Type the get_db function

JosephMart commented on services/exercises/project/db/domain/nodes/exercise.py:
```diff
@@ -0,0 +1,39 @@
+        """Serialize an `Exercise` object into a `dict`"""
+        return {
+            'id': self.id,
+            'question': self.question,
```
There is also also answer and hint

tristan957 replied:
is answer different from solution?

JosephMart commented on services/exercises/project/db/domain/nodes/topic.py:
```diff
@@ -0,0 +1,35 @@
+
+    def __init__(self, id: Optional[int, None], title: str, created_at=datetime.now(), updated_at=datetime.now()):
+        Entity.__init__(self, id)
+        self.title = title
```
name > title

JosephMart commented on services/exercises/project/db/domain/nodes/textbook.py:
```diff
@@ -0,0 +1,35 @@
+        """Serialize a `Textbook` object into a `dict`"""
+        return {
+            'id': self.id,
+            'title': self.title,
```
name > title

tristan957 replied:
I'll make the change, but I think title is better for textbooks.

JosephMart commented on services/exercises/project/db/domain/nodes/user.py:
```diff
@@ -1,7 +1,7 @@
+from typing import Dict, Union, Optional
 
 
 class User(Entity):
```
Is this supposed to duplicate users? I don't think this needs to exists in the graph db. I think the users_ids from the auth service should be stored on the textbook

tristan957 replied:
hmm that is a better idea I think. I will change the model appropriately

tristan957 commented:
Should title be changed to name for chapter as well?

JosephMart commented:
yes

JosephMart commented:
actually, i guess you can leave them all title

JosephMart commented:
what does the payload look like?

tristan957 committed:
Added hint property

tristan957 committed:
Changed title to name in Topic

tristan957 commented:
I have been working on `GET /textbooks?user_id=X`. I want the payload to look something like
```json
{
  "textbooks": [
    {
      "id": Y,
      "chapters": [
        {
          "id": X,
          "name": "MYMA",
          "children": []
        }
      ]
    }
  ]
}
```

tristan957 committed:
Revert "Changed title to name in Topic"

This reverts commit 7c56385880137b4009a511f5ef4f8ff6b1b90d03.

tristan957 committed:
Add author_id from users service

tristan957 commented:
Should a textbook have the ability for mulitple authors?

tristan957 committed:
Don't duplicate users

JosephMart commented:
Why do chapters have children key and not the came chapter key? I assume title will be in the textbooks object as well? Authors is not a thing, just user that have access to a book. Many users can have access to a textbook.

JosephMart commented:
Can the textbooks payload just be textbooks with no chapter info?

Fetching that info should be a separate payload/endpoint

tristan957 commented:
Pretty sure your example JSON in slack had `children`. I will change it to chapters. I will also restrict this endpoint to only return the textbook props

JosephMart commented:
I am saying they should be the same. both should be children or both should be chapters. They are the samething, why call it two different things

JosephMart commented:
I prefer calling it children and then exercises having an isExercise flag

tristan957 committed:
Create Textbook repo

tristan957 committed:
Update type hints

tristan957 committed:
Add textbook blueprints

tristan957 committed:
Create /textbooks endpoint

tristan957 committed:
Create package

tristan957 commented:
For the record, I am not a fan of this children approach with the `isExercise` flag.

tristan957 committed:
Catch a ValueError if user_id can't be parsed

JosephMart committed:
progress
